File: go/state.go
package main

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

const defaultStateDir = ".local/state/task"
const stateFile = "current_task"

type CurrentTask struct {
	StartTime  int64  // unix timestamp
	Name       string // task body
	FilePath   string
	LineNumber int
}

// resolveStateDir returns the state directory, using the provided override
// or falling back to ~/.local/state/task.
func resolveStateDir(override string) string {
	if override != "" {
		return expandHome(override)
	}
	home, _ := os.UserHomeDir()
	return filepath.Join(home, defaultStateDir)
}

func statePathFor(stateDir string) string {
	return filepath.Join(resolveStateDir(stateDir), stateFile)
}

// Backward-compatible: uses default state dir
func statePath() string {
	return statePathFor("")
}

func ReadCurrentTask() (*CurrentTask, error) {
	return ReadCurrentTaskFrom("")
}

func ReadCurrentTaskFrom(stateDir string) (*CurrentTask, error) {
	data, err := os.ReadFile(statePathFor(stateDir))
	if err != nil {
		if os.IsNotExist(err) {
			return nil, nil
		}
		return nil, err
	}
	line := strings.TrimRight(string(data), "\n\r")
	parts := strings.SplitN(line, "\t", 4)
	if len(parts) < 4 {
		return nil, fmt.Errorf("malformed current_task: %q", line)
	}
	ts, err := strconv.ParseInt(parts[0], 10, 64)
	if err != nil {
		return nil, fmt.Errorf("bad timestamp: %w", err)
	}
	ln, err := strconv.Atoi(parts[3])
	if err != nil {
		return nil, fmt.Errorf("bad line number: %w", err)
	}
	return &CurrentTask{
		StartTime:  ts,
		Name:       parts[1],
		FilePath:   parts[2],
		LineNumber: ln,
	}, nil
}

func WriteCurrentTask(ct CurrentTask) error {
	return WriteCurrentTaskTo("", ct)
}

func WriteCurrentTaskTo(stateDir string, ct CurrentTask) error {
	dir := filepath.Dir(statePathFor(stateDir))
	if err := os.MkdirAll(dir, 0755); err != nil {
		return err
	}
	line := fmt.Sprintf("%d\t%s\t%s\t%d\n", ct.StartTime, ct.Name, ct.FilePath, ct.LineNumber)
	return os.WriteFile(statePathFor(stateDir), []byte(line), 0644)
}

func ClearCurrentTask() error {
	return ClearCurrentTaskFrom("")
}

func ClearCurrentTaskFrom(stateDir string) error {
	err := os.Remove(statePathFor(stateDir))
	if os.IsNotExist(err) {
		return nil
	}
	return err
}

func FormatMarker(kind string, now time.Time) string {
	return fmt.Sprintf("::%-s [[%s]] %s ", kind, now.Format("2006-01-02"), now.Format("15:04"))
}

File: lua/taskbuffer/state.lua
-- state.lua — current_task state file + canonical marker formatter
-- (port of go/state.go).
--
-- These are pure primitives: the state directory is passed in (the actions /
-- config layer resolves it from config.values.state_dir and ~-expansion). They
-- do NOT read config or the wall clock.
--
-- The current_task file format is BYTE-IDENTICAL to Go's (state.go:83): a single
-- TSV line `<start_time>\t<name>\t<filepath>\t<linenumber>\n` (with trailing
-- newline). This file is consumed by start/stop logic and must not change.

local strftime = require("taskbuffer.strftime")

local M = {}

local STATE_FILE = "current_task"

---@class CurrentTask
---@field start_time integer  -- unix seconds
---@field name string         -- task body
---@field filepath string
---@field linenumber integer

-- statePathFor (state.go:32). `state_dir` is already ~-expanded by the config
-- layer; here we just join the state filename onto it.
---@param state_dir string
---@return string
function M.state_path(state_dir)
    return vim.fs.joinpath(state_dir, STATE_FILE)
end

-- SplitN-equivalent on TAB into at most `n` parts, the last part being the
-- unsplit remainder (mirrors Go's strings.SplitN(line, "\t", 4)).
local function splitn_tab(s, n)
    local parts = {}
    local start = 1
    while #parts < n - 1 do
        local i = s:find("\t", start, true)
        if not i then
            break
        end
        parts[#parts + 1] = s:sub(start, i - 1)
        start = i + 1
    end
    parts[#parts + 1] = s:sub(start)
    return parts
end

-- ReadCurrentTaskFrom (state.go:45). Missing file -> nil,nil. Malformed -> nil,err.
---@param state_dir string
---@return CurrentTask|nil, string|nil
function M.read_current_task(state_dir)
    local path = M.state_path(state_dir)
    local stat = vim.uv.fs_stat(path)
    if not stat then
        return nil, nil -- no file (os.IsNotExist -> nil,nil)
    end
    local f = io.open(path, "rb")
    if not f then
        return nil, "reading " .. path
    end
    local data = f:read("*a") or ""
    f:close()

    -- TrimRight(data, "\n\r") (state.go:53).
    local line = (data:gsub("[\n\r]+$", ""))
    local parts = splitn_tab(line, 4)
    if #parts < 4 then
        return nil, string.format("malformed current_task: %q", line)
    end
    local ts = tonumber(parts[1])
    if not ts then
        return nil, "bad timestamp: " .. parts[1]
    end
    local ln = tonumber(parts[4])
    if not ln then
        return nil, "bad line number: " .. parts[4]
    end
    return {
        start_time = math.floor(ts),
        name = parts[2],
        filepath = parts[3],
        linenumber = math.floor(ln),
    }
end

-- WriteCurrentTaskTo (state.go:78). mkdir -p the state dir, then write the TSV
-- line with a trailing newline.
---@param state_dir string
---@param ct CurrentTask
---@return boolean ok, string|nil err
function M.write_current_task(state_dir, ct)
    vim.fn.mkdir(state_dir, "p")
    local line = string.format("%d\t%s\t%s\t%d\n", ct.start_time, ct.name, ct.filepath, ct.linenumber)
    local f = io.open(M.state_path(state_dir), "wb")
    if not f then
        return false, "writing " .. M.state_path(state_dir)
    end
    f:write(line)
    f:close()
    return true
end

-- ClearCurrentTaskFrom (state.go:91). Remove the file; missing -> ok (no error).
---@param state_dir string
---@return boolean ok, string|nil err
function M.clear_current_task(state_dir)
    local path = M.state_path(state_dir)
    if not vim.uv.fs_stat(path) then
        return true -- already gone (os.IsNotExist -> nil)
    end
    local ok, err = os.remove(path)
    if not ok then
        return false, err
    end
    return true
end

-- FormatMarker (state.go:99). Canonical marker formatter. Produces
-- `<prefix><kind> [[DATE]] TIME ` (note the TRAILING space).
--
-- `now_epoch` is injected (no internal clock read) for deterministic tests.
-- `ctx` is a plain table read for { date_fmt, time_fmt, marker_prefix }. Per
-- overview D8 the CONFIGURED marker_prefix is used (Go hardcoded "::"); the
-- default "::" is byte-identical to Go.
---@param kind string
---@param now_epoch integer
---@param ctx { date_fmt:string, time_fmt:string, marker_prefix:string }
---@return string
function M.format_marker(kind, now_epoch, ctx)
    local prefix = ctx.marker_prefix or "::"
    local date = strftime.format_epoch(now_epoch, ctx.date_fmt)
    local time = strftime.format_epoch(now_epoch, ctx.time_fmt)
    return prefix .. kind .. " [[" .. date .. "]] " .. time .. " "
end

return M
